src/plugins/data/public/search/search_service.test.ts

```diff
@@ -49,26 +54,168 @@ describe('Search service', () => {
+      });
+
+      it('no notifications', () => {
+        const responder = inspector.adapter.start('request1');
+        shards = { total: 4, successful: 4, skipped: 0, failed: 0 };
+        responder.ok(getMockResponseWithShards(shards));
+        data.showWarnings(inspector.adapter, callback);
+
+        expect(notifications.toasts.addWarning).toBeCalledTimes(0);
+      });
+
+      it('notifications if no callback is provided', () => {
+        const responder = inspector.adapter.start('request1');
+        responder.ok(getMockResponseWithShards(shards));
+        data.showWarnings(inspector.adapter);
+
+        expect(notifications.toasts.addWarning).toBeCalledTimes(1);
+        expect(notifications.toasts.addWarning).toBeCalledWith({
+          title: '2 of 4 shards failed',
+          text: expect.any(Function),
+        });
+      });
+
+      it('no notifications when all warnings are filtered out', () => {
+        callback = () => true;
+        const responder = inspector.adapter.start('request1');
+        responder.ok(getMockResponseWithShards(shards));
+        data.showWarnings(inspector.adapter, callback);
+
+        expect(notifications.toasts.addWarning).toBeCalledTimes(0);
+      });
+
+      it('able to single notification when some warnings are filtered', () => {
+        callback = (warning) => warning.type === 'illegal_argument_exception';
+        shards.failures = [
+          {
+            reason: {
+              type: 'illegal_argument_exception',
+              reason: 'reason of "illegal_argument_exception"',
+            },
+          },
+          {
+            reason: {
+              type: 'other_kind_of_exception',
+              reason: 'reason of other_kind_of_exception',
+            },
+          },
+          { reason: { type: 'fatal_warning', reason: 'this is a fatal warning message' } },
+        ] as unknown as estypes.ShardFailure[];
+
+        const responder = inspector.adapter.start('request1');
+        responder.ok(getMockResponseWithShards(shards));
+        data.showWarnings(inspector.adapter, callback);
+
+        expect(notifications.toasts.addWarning).toBeCalledTimes(1);
+        expect(notifications.toasts.addWarning).toBeCalledWith({
+          title: '2 of 4 shards failed',
+          text: expect.any(Function),
+        });
+      });
+
+      it('timed_out warning', () => {
+        const responder = inspector.adapter.start('request1');
+        const response1 = getMockResponseWithShards(shards);
+        response1.json.rawResponse.timed_out = true;
+        responder.ok(response1);
+        data.showWarnings(inspector.adapter, callback);
+
+        expect(notifications.toasts.addWarning).toBeCalledTimes(1);
+        expect(notifications.toasts.addWarning).toBeCalledWith({
+          title: 'Data might be incomplete because your request timed out',
+        });
+      });
+
+      it('multiple warning notifications when multiple responses have shard failures', () => {
```

Comment:
There should be 1 warning per end-to-end logic block. The shard failure modal should be able to display multiple responses since we work with batched requests.